### ISSUE TITLE
Latest version conflicts with WP Bakery Visual Builder and/or WooCommerce

### ISSUE THREAD

**Issue by janboddez** [semperfiwebdesign/all-in-one-seo-pack#2778]
Alright, so I've got this page that uses a Visual Builder layout. First row essentially runs a WooCommerce-related database query, then displays the first three results.

With All-in-One SEO (v3.2.1) enabled, the page title (which should be hidden) is displayed, but not before it is wrongly set to that of the first product, i.e., the first query result. The rest of the Visual Builder layout is simply never rendered. (Essentially, the entire page is pretty much messed up/gone.)

With All-in-One SEO disabled, everything returns to normal (except for, well, AIOSEOP-related goodies being gone).

Enabled WordPress debugging, but there's no clear PHP errors or anything. Seems like a certain function of filter that this plugin hooks into doesn't quite return the correct value? (Maybe echoes it instead?) Could be the same happens with other custom post type queries, haven't been able to check, yet.

Happens on PHP 7.2 and 7.3, WordPress 5.2.2, WooCommerce 3.6.5, in combination with 3.2.1. Never noticed anything with older versions. I'll gladly provide more information if you tell me what I should be looking at.

**Comment by wpsmort** [comment]
@janboddez The first thing to do is to remove the WooCommerce-related database query and see if everything works correctly.  If it does, then  you know the problem is with that query.  

It sounds like your query is fetching product IDs and storing them so that when All in One SEO runs its query to fetch the page title, it's getting the product ID stored by your query.

**Comment by janboddez** [comment]
The behavior is definitely gone with WP Bakery VB disabled, i.e., when the custom products queries aren't run.

I also tried reproducing the error on 3.2, and it's there. With AIOSEOP v3.1.1, all is good again.

I haven't been able to set up a test site, let alone try and fix the root cause, but I wouldn't be surprised if it has to do with fetching the title from global `$post` rather than, e.g., `get_queried_object_id()`.

In that sense, the behavior I'm describing seems quite similar to what's mentioned in #2729. Of note: all of this happens on a static homepage with custom (via Visual Builder, that is, I unfortunately didn't write these) WC product queries being called inside it. Other, simpler static pages are unaffected.

As to why VB seems to then choke and give up altogether, I wouldn't know ...

Will get back when I've got more info.

**Comment by wpsmort** [comment]
Odd that you can't reproduce the problem with v3.1.1.  That issue you referenced exists in v3.1.1 and older versions.  

Is the homepage also set as the Shop page?  We're aware of an issue where we're incorrectly getting the ID of the latest product rather than the ID of the page when the Shop page is set as the static front page.

**Comment by janboddez** [comment]
It's not the shop page, either. It's a regular page that runs two additional queries to display the 3 newest products, and 3 random products from a specific category. The first item of that second query seems to be the one wrongly picked up as the 'current page' (and things sort of get screwed up from there on).

**Comment by wpsmort** [comment]
Then I'd say you've narrowed the problem down to the second query.  If it doesn't happen with the first query but does with the second query then there could be something wrong in that query.

**Comment by janboddez** [comment]
I'm able to work around the issue, I found, by disabling Schema.org markup. (Enabling this option on versions before 3.2 does not trigger the behavior.)

~~Also, the problem does not occur for each and every Visual Builder element, but it definitely occurs whenever their Products element (running said Products query) is present.~~ In fact, these elements (shortcodes, actually) are part of the theme on this site: Laborator's Oxygen. Visual Builder itself may thus not at all be what triggers this.

I actually just noticed that commenting out `$post_description = aioseop_do_shortcodes( $post_description );` on line 173 of `inc/schema/graphs/graph-webpage.php also works around the issue`.

And not doing so leads to part of the page's HTML being echoed *inside* the Schema.org JSON, which can't be right. Seems `aioseop_do_shortcodes` unintentionally leads to markup being printed rather than returned.

Long story short: the site's theme may very well be to blame, but still, something in the parsing of Schema.org inputs seems to have changed between versions 3.1 and 3.2, bringing this to light.

**Comment by wpsmort** [comment]
Thanks @janboddez.

@EkoJR Is this the same problem you're working on?  If not, can you take a look at this in conjunction with the problem reported in #2774.

**Comment by EkoJR** [comment]
@wpsmort Yea, it does sound like another shortcode conflict that would need to be added to the `$conflicting_shortcodes` array.

Another way to workaround this is to add an SEO Description or Excerpt to the post/page. So that it doesn't try to use the post_content to generate a description.

**Comment by wpsmort** [comment]
@janboddez We have a potential fix for this.  Please can you test it by downloading the beta code from here - https://github.com/semperfiwebdesign/all-in-one-seo-pack/tree/3.2.2

Click the Clone or Download button and then download as a Zip file and install it on a test site if possible.  Make sure that Use Schema.org is set to Enabled under All in One SEO > General Settings > Schema Settings.

Please let us know if this resolves this issue or not.

**Comment by janboddez** [comment]
Tested on two different setups (live and development) that previously showed the problem, and I can confirm it is gone with this version.

**Comment by michaeltorbert** [comment]
@janboddez Awesome! Thanks for lettings us know. :)
